### ISSUE TITLE
split dns with systemd-resolved + tailscale stop working after a deploy

### ISSUE THREAD

**Issue by bbigras** [NixOS/nixpkgs#104414]
**Describe the bug**

I have a tunnel to work using tailscale.

I'm using systemd-resolved to resolve work-related domains over the tailscale tunnel.

When I redeploy with `nixos-rebuild switch` my work-related domains doesn't resolve anymore.

I'm thinking maybe tailscale replaces my dns settings.

I think I tried `tailscale up -accept-dns false` without success.

EDIT: my "main dns config" seems to still work since https://www.cloudflare.com/ssl/encrypted-sni/ tells me that I'm still using dnssec with tls.

**To Reproduce**
1. ...
2. ...
3. ...

**Expected behavior**

**Screenshots**

**Additional context**

**Notify maintainers**

@danderson @mbaillie

**Metadata**

nix-pre-commit-hooks: hooks up to date
 - system: `"x86_64-linux"`
 - host os: `Linux 5.7.10-zen1, NixOS, 21.03pre-git (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.8`
 - channels(root): `"nixos-21.03pre251123.b839d4a8557"`
 - channels(bbigras): `"home-manager"`
 - nixpkgs: `/run/current-system/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by flokli** [comment]
Can you extend the description with some more bits of your configuration, so this can be reproduced?

**Comment by danielphan2003** [comment]
I think this has to do with a race condition where Tailscale starts *after* systemd-resolved, which makes Tailscale DNS unreachable by the time systemd-resolved wanted to use it as default, but instead fall back to whatever DNS your system may have. In your case, however, there might have been none.

A solution would be to just restart systemd-resolved whenever Tailscale starts, but that seems hacky. After all, it should have worked out-of-the-box, but with more complex [usage][danielphan2003-networkd-config] like I have, maybe using that [hack][tailscale-resolved-race-condition-hack] is reasonable enough. See [this][tailscale-overlay] for more info.

[danielphan2003-networkd-config]: https://github.com/danielphan2003/flk/blob/main/nixos/profiles/network/networkd/default.nix
[tailscale-resolved-race-condition-hack]: https://github.com/danielphan2003/flk/blob/main/nixos/profiles/network/tailscale/default.nix#L33
[tailscale-overlay]: https://github.com/danielphan2003/flk/blob/main/overlays/tailscale.nix

**Comment by flokli** [comment]
@danderson could you take a look into this?

I'm not entirely sure what tailscale is doing to configure itself in resolved setups (and to ensure it stays configured the way it is, both after networkd restarts and tailscale restarts).

If it's only editing `/etc/hosts`, it won't work, as resolved brings its own NSS module.

**Comment by danderson** [comment]
This was likely fixed in either 1.22 or 1.24: we used to not detect systemd-resolved restarting (which clears all its DNS config), but some months ago we added the logic to monitor dbus and reprogram resolved after a restart. Anecdotally, my NixOS desktops and servers don't have any Tailscale DNS issues across deploys at this point.

That fix also addressed a more subtle race condition during bootup, where tailscaled could sometimes start up prior to resolved being registered on the system bus. When that happened, tailscaled would switch to direct /etc/resolv.conf management, which broke a bunch of things. The change to monitor the bus for resolved also fixed that, by having tailscaled figure out that resolved _should_ be in charge, and waiting for it to appear on the system bus.

IOW, unless this is still reproducible, I think this can be closed as fixed.
